### PR TITLE
Works on arm v3 minimalish

### PR DESCRIPTION
**Pull request by holdenk**
## Why are these changes needed?

ARM is becoming increasingly popular with both the M1 and graviton processors. I myself run a mixed cluster of ARM64 (jetson, rpis, etc.) & AMD64 machines  (more traditional).

## Related issue number

Related to the various request for ARM wheels but this is just the docker image side of it.

## Checks

- [ X ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ X ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ X ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ X ] Release tests (`builder-docker.sh`)
   - [ ] This PR is not tested :(
 
You can take a look at the containers that are built under: https://hub.docker.com/u/holdenk (all with the extra `ray-` prefix).

### CODE REVIEW

**Comment by richardliaw** [comment]
this is awesome! 

@wuisawesome can you help shepherd this?

cc @amholler @simonfall

**Comment by holdenk** [comment]
Hey y'all is this something y'all would be interested in? I'm happy to update it but if not I'll just keep my branch as-is so I can have a stable base to work on my book from.

**Comment by wuisawesome** [comment]
Hey, sorry about that. I lost this PR. I'll give this PR a review, and in general, I think we're pretty open to ARM support, that being said, this we will almost certainly, inevetably, break ARM support again if we don't get the CI set up for it.

**Comment by wuisawesome** [comment]
@simon-mo do you know how hard it would be to set up an ARM build on graviton instances?

File: python/requirements.txt [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #
```
Comment (wuisawesome) [inline python/requirements.txt:6]: Can we avoid deleting things from this file? In order to get the arm build to work, I believe we should only have to change the requirements for the wheel which are in `setup.py` (under `setup_spec.install_requires`)

File: python/requirements.txt [hunk 2]
```diff
@@ -6,6 +6,7 @@
...
 # setup.py install_requires
 aiohttp==3.7
+aiohttp_cors
```
Comment (wuisawesome) [inline python/requirements.txt:9]: Do we need this? And do you. mind dropping a comment explaining why if we do?

File: python/ray/autoscaler/kubernetes/operator_configs/operator.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
+operator_role:
```
Comment (wuisawesome) [inline python/ray/autoscaler/kubernetes/operator_configs/operator.yaml:1]: Is this necessary for making the ARM build work?

File: docker/ray/Dockerfile [hunk 4]
```diff
@@ -1,11 +1,23 @@
 ARG GPU
-FROM rayproject/ray-deps:nightly"$GPU"
-ARG WHEEL_PATH
-ARG FIND_LINKS_PATH=".whl"
+ARG DOCKER_PREFIX
```
Comment (wuisawesome) [inline docker/ray/Dockerfile:2]: Is there a way to keep these docker file changes a bit more minimal? 

Also @ijrsvt do you mind taking a look at this part?

File: docker/ray/Dockerfile [hunk 5]
```diff
@@ -1,11 +1,23 @@
...
-RUN $HOME/anaconda3/bin/pip --no-cache-dir install `basename $WHEEL_PATH`[all] \
---find-links $FIND_LINKS_PATH && sudo rm `basename $WHEEL_PATH`
+# Specify bazel (see https://github.com/bazelbuild/bazel/issues/12887)
+ENV USE_BAZEL_VERSION=3.7.2
+RUN sudo apt-get update && sudo apt-get install -y curl bash
+RUN echo "export PATH=~/anaconda3/bin/:~/.cargo/bin:\$PATH" >> ~/.bashrc
+SHELL ["/bin/bash", "-c"]
+RUN export PATH=$HOME/.cargo/bin:$HOME/anaconda3/bin/:$PATH
+# Install py-spy no arm on conda yet so fall through to pypi wheel
+RUN (mamba install --yes py-spy) || (pip install --upgrade pip && pip install py-spy)
```
Comment (wuisawesome) [inline docker/ray/Dockerfile:20]: Can we stick to either pulling our dependencies from pypi or just directly installing them from github (ideally pinned to a tagged commit)? It just makes it a bit easier for us to track where we're pulling dependencies in from?

File: deploy/components/example_cluster.yaml [hunk 6]
```diff
@@ -73,6 +73,7 @@ spec:
             requests:
               cpu: 1000m
               memory: 512Mi
+              ephemeral-storage: 1Gi
```
Comment (wuisawesome) [inline deploy/components/example_cluster.yaml:76]: hmmm do we need to add ephemeral storage/is there a benefit to it? If not, can we revert the changes to this file for the sake of keeping the PR minimal?

File: docker/base-deps/Dockerfile [hunk 7]
```diff
@@ -27,53 +27,64 @@ RUN apt-get update -y \
...
+	--quiet "https://github.com/conda-forge/miniforge/releases/download/4.8.5-1/Miniforge3-4.8.5-1-Linux-${arch}.sh" \
+	-O /tmp/miniconda.sh \
     && /bin/bash /tmp/miniconda.sh -b -u -p $HOME/anaconda3 \
-    && $HOME/anaconda3/bin/conda init \ 
+    && $HOME/anaconda3/bin/conda init \
     && echo 'export PATH=$HOME/anaconda3/bin:$PATH' >> /home/ray/.bashrc \
     && rm /tmp/miniconda.sh \
-    && $HOME/anaconda3/bin/conda install -y \
-        libgcc python=$PYTHON_VERSION \
+    && $HOME/anaconda3/bin/conda install -y mamba \
```
Comment (wuisawesome) [inline docker/base-deps/Dockerfile:60]: Can we stick to `pip` and `apt-get` for dependencies? They're admittedly old and boring, but reliable and well understood.

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Sorry for the late review. I left a couple of comments. As a general theme, is it possible to distill this to the minimal set of things needed to get ARM to work?

File: docker/ray-ml/Dockerfile [hunk 8]
```diff
@@ -1,27 +1,29 @@
 ARG GPU
-FROM rayproject/ray:nightly"$GPU"
+ARG DOCKER_PREFIX
```
Comment (ijrsvt) [inline docker/ray-ml/Dockerfile:2]: Could `DOCKER_PREFIX` be defaulted to `rayproject`?

File: docker/ray-ml/Dockerfile [hunk 9]
```diff
@@ -1,27 +1,29 @@
...
+
+FROM "$DOCKER_PREFIX"ray:nightly"$GPU"
 ARG PYTHON_MINOR_VERSION=7
 
 # We have to uninstall wrapt this way for Tensorflow compatibility
 COPY requirements.txt ./
 COPY requirements_ml_docker.txt ./
 COPY requirements_rllib.txt ./
-COPY requirements_tune.txt ./requirements_tune.txt
+# COPY requirements_tune.txt ./requirements_tune.txt
```
Comment (ijrsvt) [inline docker/ray-ml/Dockerfile:11]: Is this intentional?

File: docker/ray/Dockerfile [hunk 10]
```diff
@@ -1,11 +1,23 @@
 ARG GPU
-FROM rayproject/ray-deps:nightly"$GPU"
-ARG WHEEL_PATH
-ARG FIND_LINKS_PATH=".whl"
+ARG DOCKER_PREFIX
+
+# Pull the wheel from the ray-wheel build
+FROM "$DOCKER_PREFIX"ray-wheel:nightly"$GPU" AS raywheel
+WORKDIR /ray
```
Comment (ijrsvt) [inline docker/ray/Dockerfile:6]: What is the reason for switching from COPY -> from another image?

**Review by ijrsvt — COMMENTED** [review]
Quick question, what are the `ray-nb` and `ray-wheel` images for?

**Comment by holdenk** [comment]
> Quick question, what are the `ray-nb` and `ray-wheel` images for? 

So ray-wheel builds the ray wheel, this allows us to build for multiple archs and correctly pull it in provided that one either has a remote docker arm context or qemu setup.

ray-nb is a Jupyter notebook container, I can drop it from this PR.

File: docker/ray/Dockerfile (hunk 5 above)
Comment (holdenk) [reply docker/ray/Dockerfile:20]: Yeah I can change this one to just always use pypi

File: deploy/components/example_cluster.yaml (hunk 6 above)
Comment (holdenk) [reply deploy/components/example_cluster.yaml:76]: True unrelated (I need it to deploy on my cluster but I can do it as a separate PR).

File: docker/base-deps/Dockerfile (hunk 7 above)
Comment (holdenk) [reply docker/base-deps/Dockerfile:60]: Mamba is pretty common these days and it gives much better error messages when it can't resolve the packages (more likely to happen with ARM because less conda packages published).

That being said it is optional, eg the PR would work without it just more painful to debug long term.

File: python/requirements.txt (hunk 1 above)
Comment (holdenk) [reply python/requirements.txt:6]: requirements.txt isn't quite up to date with setup.py so this gets them more in sync. Part of this is because we install the wheel with no-index which means we assume all of the dependencies are installed, I change that part though but I still think having a requirements.txt that is installable on ARM is useful.

File: python/requirements.txt (hunk 2 above)
Comment (holdenk) [reply python/requirements.txt:9]: We do in that it's already required in `setup.py` people just haven't been following the instructions in setup.py to keep requirements.txt up to date.

File: docker/ray/Dockerfile (hunk 5 above)
Comment (holdenk) [reply docker/ray/Dockerfile:20]: To be fair though, just using pypi is going to be slower because some of the packages can be installed faster from conda-forge.

File: docker/ray/Dockerfile (hunk 10 above)
Comment (holdenk) [reply docker/ray/Dockerfile:6]: So we copy from another image so that we get the right wheel since the raywheel image contains the wheel built for our arch.

File: docker/ray-ml/Dockerfile (hunk 8 above)
Comment (holdenk) [reply docker/ray-ml/Dockerfile]: Sure :)

File: docker/ray-ml/Dockerfile (hunk 9 above)
Comment (holdenk) [reply docker/ray-ml/Dockerfile:10]: Probably, but I don't remember at this point so I'll re-enable it and figure out what it broke :)

File: docker/base-deps/Dockerfile (hunk 7 above)
Comment (holdenk) [reply docker/base-deps/Dockerfile:61]: Oh also, just to be clear the existing Dockerfiles don't just use apt & pip, they already use conda, mamba pulls from conda-forge so it's not really a new source.

File: python/ray/autoscaler/kubernetes/operator_configs/operator.yaml (hunk 3 above)
Comment (holdenk) [reply python/ray/autoscaler/kubernetes/operator_configs/operator.yaml:1]: No it is not, sorry this is an unrelated K8s change, I'll move it over.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.
